Add support for overlay networking and GCP

benmoss opened the pull request:
Part of https://github.com/cloudfoundry-incubator/kubo-release/pull/320

Creating the virtual network interface on Windows has the effect of breaking the route to the 169.254.169.254 metadata serer. We just have to recreate that route after we bring up flannel to make the GCP cloud provider work. Other stuff is as similar as possible to Linux, we're using `kubectl proxy` to get TLS rather than doing it with `curl` since using custom certs via `--cacert` apparently doesn't work easily on Windows.

benmoss committed:
Initial support for Windows overlay networking

benmoss committed:
Add support for GCP

benmoss commented:
Important to note that this functionality only works on k8s 1.14 or higher

benmoss commented:
@tvs @zankich

zankich commented:
@benmoss the tracker integration wasn't configured for this repo, so I added a story to track this work https://www.pivotaltracker.com/story/show/165729634

benmoss committed:
Timeout flannel post-start on failure

Otherwise the bosh deploy will hang forever

benmoss committed:
Wait for kubectl proxy to be up before patching network status

benmoss committed:
Disable address management when running on GCP

https://github.com/kubernetes/kubernetes/issues/75561

benmoss committed:
Add retries to kube-proxy prestart check

benmoss committed:
Manually re-add gateway to routing table

On Azure we are seeing VMs often lose their gateway after the switch to
HNS networking is complete. We couldn't figure out why so this is a
reasonable workaround. It will just log an error and continue in the
case of the route already existing.

Commit:
Restart docker after reconfiguring gateway

Without this `docker pull`s seemed to just hang forever

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Use BOSH agent UUID as hostname on Azure

BOSH Windows doesn't set this to the hostname, so we need to use an
override

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Remove misleading comment

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Set hostname-override on kube-proxy when running on Azure

Need to override it there too so it matches kubelet

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Conditional networking setup for overlay

Matching the logic of https://github.com/microsoft/SDN/blob/3414f2d3f036a4e95187c4ae01e6b2db0df78484/Kubernetes/windows/helper.v2.psm1#L651-L726

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Bump flannel to latest master

Includes changes from https://github.com/coreos/flannel/pull/1137 that
fix NodePort connectivity

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Add fix for GCP persistent volumes

Sourced from https://github.com/pjh/gce-tools, this DLL allows kubelet
to run the `Get-GcePdName` powershell commandlet.

https://github.com/kubernetes/kubernetes/issues/74674

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Uses the correct hostname for AWS nodes

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Removes any existing HNSNetworks before applying overlay / host gateway
- When upgrading from host gateway to overlay, this will ensure the
flannel-windows job does not fail

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Updates cni plugins. Updates cni.conf to include DNS capability

Signed-off-by: Ben Moss <bmoss@pivotal.io>

Commit:
Add cniVersion to CNI config

 It's now required due to
 https://github.com/kubernetes/kubernetes/pull/80482 adding CNI
 validation to the kubelet

Signed-off-by: Ben Moss <bmoss@pivotal.io>
